vxtelegram/telegram.py

```diff
@@ -157,31 +185,104 @@ def handle_outbound_message(self, message):
         url = self.get_outbound_url('sendMessage')
         http_client = HTTPClient(self.agent_factory())
 
-        try:
-            r = yield http_client.post(
-                url=url,
-                data=json.dumps(outbound_msg),
-                headers={'Content-Type': ['application/json']},
-            )
+        r = yield http_client.post(
+            url=url,
+            data=json.dumps(outbound_msg),
+            headers={'Content-Type': ['application/json']},
+            allow_redirects=False,
+        )
+
         # Telegram redirects our request if our bot token is invalid
-        except ResponseFailed as e:
-            yield self.outbound_failure(message_id,
-                                        'Invalid token (redirected)\n%s' % e)
+        if r.code == http.FOUND:
+            yield self.outbound_failure(
+                message_id=message_id,
+                message='Invalid bot token (redirected)',
+                error='Redirected',
+            )
             return
+
         try:
             res = yield r.json()
-        except ValueError as e:
-            yield self.outbound_failure(message_id,
-                                        'Expected JSON response\n%s' % e)
+        except ValueError:
+            yield self.outbound_failure(
+                message_id=message_id,
+                message='Expected JSON response',
+                error='error code %s' % r.code,
+            )
             return
 
         if r.code == http.OK and res['ok']:
-            yield self.publish_ack(user_message_id=message_id,
-                                   sent_message_id=message_id)
+            yield self.outbound_success(message_id)
         else:
-            yield self.outbound_failure(message_id, res['description'])
+            yield self.outbound_failure(
+                message_id=message_id,
+                message='Bad response from Telegram',
+                error=res['description'],
+            )
 
     @inlineCallbacks
-    def outbound_failure(self, message_id, reason):
-        yield self.publish_nack(message_id, 'Failed to send message: %s' %
-                                reason)
+    def outbound_failure(self, message_id, message, error):
+        yield self.publish_nack(message_id, '%s: %s' % (message, error))
+        yield self.add_status_bad_outbound(message, error)
+
+    @inlineCallbacks
+    def outbound_success(self, message_id):
+        yield self.publish_ack(message_id, message_id)
+        yield self.add_status_good_outbound()
+
+    def add_status_bad_inbound(self, message, error=None):
+        details = {}
+        if error is not None:
+            details.update({'error': error})
+        return self.add_status(
+            status='down',
+            component='telegram_inbound',
+            type='bad_inbound',
```

Comment:
I think it would be helpful to be able to use the `type` field to differentiate between the different types of bad inbound cases (redirects, json parse errors and api error responses), for example, so that the application handling the status message can handle each error differently.